internal/component/azure/servicebus/subscription.go

```diff
@@ -204,69 +313,60 @@ func (s *Subscription) Close(closeCtx context.Context) {
 	s.cancel()
 }
 
-func (s *Subscription) handleAsync(ctx context.Context, msg *azservicebus.ReceivedMessage, handler HandlerFunc) {
+// handleAsync handles messages from azure service bus asynchronously.
+// runHandlerFn is responsible for calling the message handler function
+// and marking messages as complete/abandon.
+func (s *Subscription) handleAsync(ctx context.Context, msgs []*azservicebus.ReceivedMessage, runHandlerFn func()) {
 	go func() {
 		var (
 			consumeToken           bool
 			takenConcurrentHandler bool
-			err                    error
 		)
 
 		defer func() {
-			// Release a handler if needed
-			if takenConcurrentHandler {
-				<-s.handleChan
-				s.logger.Debugf("Released message handle for %s on %s", msg.MessageID, s.entity)
-			}
-
-			// If we got a retriable error (app handler returned a retriable error, or a network error while connecting to the app, etc) consume a retriable error token
-			// We do it here, after the handler has been released but before removing the active message (which would allow us to retrieve more messages)
-			if consumeToken {
-				s.logger.Debugf("Taking a retriable error token")
-				before := time.Now()
-				_ = s.retriableErrLimit.Take()
-				s.logger.Debugf("Resumed after pausing for %v", time.Now().Sub(before))
+			for _, msg := range msgs {
+				// Release a handler if needed
+				if takenConcurrentHandler {
+					<-s.handleChan
+					s.logger.Debugf("Released message handle for %s on %s", msg.MessageID, s.entity)
+				}
+
+				// If we got a retriable error (app handler returned a retriable error, or a network error while connecting to the app, etc) consume a retriable error token
+				// We do it here, after the handler has been released but before removing the active message (which would allow us to retrieve more messages)
+				if consumeToken {
+					s.logger.Debugf("Taking a retriable error token")
+					before := time.Now()
+					_ = s.retriableErrLimit.Take()
+					s.logger.Debugf("Resumed after pausing for %v", time.Since(before))
+				}
+
+				// Remove the message from the map of active ones
+				s.removeActiveMessage(msg.MessageID, *msg.SequenceNumber)
+
+				// Remove an entry from activeMessageChan to allow processing more messages
+				<-s.activeMessagesChan
```

Comment:
Here's where the issue I was mentioning before happens. You are here inside a loop for each message, and this may cause issues.